operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/CustomResourceCache.java

```diff
@@ -44,19 +44,21 @@ public CustomResourceCache(ObjectMapper objectMapper, Metrics metrics) {
     resources = metrics.monitorSizeOf(new ConcurrentHashMap<>(), "cache");
   }
 
-  public void cacheResource(CustomResource resource) {
+  @SuppressWarnings("unchecked")
+  public void cacheResource(T resource) {
     cacheResource(resource, passthrough);
   }
 
-  public void cacheResource(CustomResource resource, Predicate<CustomResource> predicate) {
+  public void cacheResource(T resource, Predicate<T> predicate) {
     try {
       lock.lock();
       final var uid = getUID(resource);
       if (predicate.test(resources.get(uid))) {
         if (passthrough != predicate) {
           log.trace("Update cache after condition is true: {}", getName(resource));
         }
-        resources.put(uid, resource);
+        // defensive copy
+        resources.put(getUID(resource), clone(resource));
```

Comment:
why are we cloning here?

Reply:
The reason is that as example, if for some reason the process that applies the finalizer fails when updating the object, then the cache would still reporting the object as having the finalizer which is not correct.

it probably won't matter so if it should be removed, I ca do it in a subsequent pr